# Données 5 sc une semaine

In [1]:
#Charger les biblio
import numpy as np
import pandas as pd
from pandas import read_csv
import matplotlib.pyplot as plt
from matplotlib import cm
import seaborn as sns
import numpy as np
import string 
from subprocess import check_output
import seaborn as sns
from pylab import *
%matplotlib inline
sns.set()

In [2]:
# Lire la table avec 'header'
df = pd.read_csv('Hmvl33.csv', sep=';', engine='python')

On sépare la table en deux

In [7]:
dff = df[0:5871318]

In [8]:
# Remplacer les NaN par des zeros dans trame_hmvl
dff.trame_hmvl.fillna('0', inplace=True)

C:\Users\oumayma.doughmi\ANACONDA\lib\site-packages\pandas\core\generic.py:6130: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


In [10]:
# Reindexer la dataframe
dff.index = range(len(dff))

In [11]:
#changeaons les noms de colonnes
dff.columns = ['code', 'id','equipement_id', 'date','trame']

In [12]:
# On enlève les @
dff["trame"] = dff["trame"].str.replace("@", "")

C:\Users\oumayma.doughmi\ANACONDA\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [13]:
# On ajoute une colonne Count pour compter chacune
dff['Count'] = dff['trame'].str.len()

C:\Users\oumayma.doughmi\ANACONDA\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [14]:
# Supprimons les lignes differentes de zero
dff = dff[dff.Count != 0]

In [15]:
# Supprimons les lignes differentes de 1
dff = dff[dff.Count != 1]

In [16]:
d = dff.groupby("Count").count()
# La somme cumulé
d['CUMS'] = d['id'].cumsum(axis = 0)

In [23]:
d[75:85]

code      id  equipement_id    date   trame     CUMS
Count                                                        
116         5       5              5       5       5  4654041
117         2       2              2       2       2  4654043
118         3       3              3       3       3  4654046
119         1       1              1       1       1  4654047
120       275     275            275     275     275  4654322
121    128650  128650         128650  128650  128650  4782972
122       708     708            708     708     708  4783680
123       644     644            644     644     644  4784324
124       411     411            411     411     411  4784735
125       209     209            209     209     209  4784944

Le nombre totale de caractère qui représente 100% est de $50.447.481$.

$5044748$ représente $95\%$ de données qui correspond à la valeur au champs 'Count'.

On considère alors, pour ces données, le nombre maximal de caractère est de 11.

In [25]:
# Supprimons la colonne
dff = dff.drop(['Count'], axis=1)

In [26]:
liste = ["x0","x1","x2","x3","x4","x5","x6","x7","x8","x9",'x10']
# Ajouter les colonnes, en séparant la colonne 'trame' 
for i in range(len(liste)):
    dff[liste[i]] = dff["trame"].str.slice(11*i,11*(i+1))

In [27]:
# Supprimer les colonnes xj
df0 = dff.drop(['x3','x1','x2','x4','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
# df1
df1 = dff.drop(['x3','x0','x2','x4','x5','x6','x7','x8', 'x9', 'x10',  'trame'], axis=1)
df2 = dff.drop(['x3','x1','x0','x4','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df3 = dff.drop(['x0','x1','x2','x4','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df4 = dff.drop(['x0','x1','x2','x3','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df5 = dff.drop(['x0','x1','x2','x4','x3','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df6 = dff.drop(['x0','x1','x2','x4','x5','x3','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df7 = dff.drop(['x0','x1','x2','x4','x5','x6','x3','x8', 'x9', 'x10', 'trame'], axis=1)
df8 = dff.drop(['x0','x1','x2','x4','x5','x6','x3','x7', 'x9', 'x10', 'trame'], axis=1)
df9 = dff.drop(['x0','x1','x2','x4','x5','x6','x3','x7', 'x8', 'x10',  'trame'], axis=1)

df10 = dff.drop(['x0','x1','x2','x4','x5','x6','x3','x7', 'x9', 'x8','trame'], axis=1)


In [28]:
liste = ["x0","x1","x2","x3","x4","x5","x6","x7","x8","x9",'x10']
liste2 = [df0,df1,df2,df3,df4,df5,df6,df7,df8,df9,df10]

    # Séparer la colonne xi en 4 colonnes  
for i, f in zip(range( len(liste)), liste2):
    f["voie"] = f[liste[i]].str.slice(0,1)
    f["horodatage"] = f[liste[i]].str.slice(1,5)
    f["vitesse"] = f[liste[i]].str.slice(5,8)
    f["longueur"] = f[liste[i]].str.slice(8,11)

In [29]:
# rename columns
for i,f in zip(range( len(liste)), liste2):
    
    f.rename(columns={liste[i]:'x'}, inplace=True)

In [ ]:
# Concatener les data
result = pd.concat(liste2)

In [ ]:
# Reindexer la dataframe
result.index = range(len(result))

In [ ]:
# Eliminer les lignes vides
df = result.dropna(axis = 0, how ='any', thresh = None, subset = None, inplace = False)

In [3]:
df.head()

code    id  equipement_id           date_reception  \
0                   MBS  7357           7357  2019-02-18 00:00:00.146   
1                   MPA  7356           7356  2019-02-18 00:00:00.146   
2                   MPH  7355           7355  2019-02-18 00:00:00.146   
3  SFB-TPCCEN_RDTSCB001  7850           7850  2019-02-18 00:00:01.374   
4  SMA-INTBAU_RDTSCB001  7859           7859  2019-02-18 00:00:01.374   

                                          trame_hmvl  
0  0543110903705431109037256930750412569307504125...  
1                                                  @  
2                 058520920440585209204405852092044@  
3                                                  @  
4                                                  @

In [3]:
# Remplacer les NaN par des zeros dans trame_hmvl
df.trame_hmvl.fillna('0', inplace=True)

In [4]:
# Reindexer la dataframe
df.index = range(len(df))

In [5]:
#changeaons les noms de colonnes
df.columns = ['code', 'id','equipement_id', 'date','trame']

In [6]:
# On enlève les @
df["trame"] = df["trame"].str.replace("@", "")

In [7]:
# On ajoute une colonne Count pour compter chacune
df['Count'] = df['trame'].str.len()

In [9]:
df.Count.max()/11

60.0

In [8]:
# Supprimons les lignes differentes de zero
df = df[df.Count != 0]

In [9]:
# Supprimons les lignes differentes de 1
df = df[df.Count != 1]

In [12]:
dff = df.groupby("Count").count()
# La somme cumulé
dff['CUMS'] = dff['id'].cumsum(axis = 0)

In [17]:
dff[80:90]

code      id  equipement_id    date   trame     CUMS
Count                                                        
117         3       3              3       3       3  9450644
118         3       3              3       3       3  9450647
119         1       1              1       1       1  9450648
120       567     567            567     567     567  9451215
121    244209  244209         244209  244209  244209  9695424
122      1370    1370           1370    1370    1370  9696794
123      1138    1138           1138    1138    1138  9697932
124       679     679            679     679     679  9698611
125       309     309            309     309     309  9698920
126        93      93             93      93      93  9699013

Le nombre totale de caractère qui représente 100% est de $10.186.691$.

$9695424$ représente $95.17\%$ de données qui correspond à la valeur au champs 'Count'.

On considère alors, pour ces données, le nombre maximal de caractère est de 11.

### il faut créer 11 nouveaux champs qui contient les 11 caractères pour chaque véhicules.

In [10]:
# Supprimons la colonne
df = df.drop(['Count'], axis=1)

In [11]:
liste = ["x0","x1","x2","x3","x4","x5","x6","x7","x8","x9",'x10']
# Ajouter les colonnes, en séparant la colonne 'trame' 
for i in range(len(liste)):
    df[liste[i]] = df["trame"].str.slice(11*i,11*(i+1))

In [42]:
# Supprimer les colonnes xj
df0 = df.drop(['x3','x1','x2','x4','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
# df1
df1 = df.drop(['x3','x0','x2','x4','x5','x6','x7','x8', 'x9', 'x10',  'trame'], axis=1)
df2 = df.drop(['x3','x1','x0','x4','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df3 = df.drop(['x0','x1','x2','x4','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df4 = df.drop(['x0','x1','x2','x3','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df5 = df.drop(['x0','x1','x2','x4','x3','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df6 = df.drop(['x0','x1','x2','x4','x5','x3','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df7 = df.drop(['x0','x1','x2','x4','x5','x6','x3','x8', 'x9', 'x10', 'trame'], axis=1)
df8 = df.drop(['x0','x1','x2','x4','x5','x6','x3','x7', 'x9', 'x10', 'trame'], axis=1)
df9 = df.drop(['x0','x1','x2','x4','x5','x6','x3','x7', 'x8', 'x10',  'trame'], axis=1)

df10 = df.drop(['x0','x1','x2','x4','x5','x6','x3','x7', 'x9', 'x8','trame'], axis=1)


MemoryError: 

In [41]:
liste = ["x0","x1","x2","x3","x4","x5","x6","x7","x8","x9",'x10']
liste2 = [df0,df1,df2,df3,df4,df5,df6,df7,df8,df9,df10]

    # Séparer la colonne xi en 4 colonnes  
for i, f in zip(range( len(liste)), liste2):
    f["voie"] = f[liste[i]].str.slice(0,1)
    f["horodatage"] = f[liste[i]].str.slice(1,5)
    f["vitesse"] = f[liste[i]].str.slice(5,8)
    f["longueur"] = f[liste[i]].str.slice(8,11)

KeyError: 'x0'

In [40]:
# rename columns
for i,f in zip(range( len(liste)), liste2):
    
    f.rename(columns={liste[i]:'x'}, inplace=True)

MemoryError: 

In [39]:
# Concatener les data
#result = df0.append([df1, df2, df3, df4, df5, df6, df7, df8, df9, df10])
result = pd.concat(liste2)

MemoryError: 

In [ ]:
# Reindexer la dataframe
result.index = range(len(result))

In [ ]:
# Eliminer les lignes vides
df = result.dropna(axis = 0, how ='any', thresh = None, subset = None, inplace = False)

In [ ]:
len(result)-len(df)

In [ ]:
#Exporter la table
df.to_csv('5sc_to.csv', index = False, encoding = 'utf-8')

In [ ]:
df = pd.read_csv('5sc_to.csv', sep=';', engine='python')

In [31]:
df7.head()

code    id  equipement_id                     date x voie horodatage  \
0  MBS  7357           7357  2019-02-18 00:00:00.146                     
2  MPH  7355           7355  2019-02-18 00:00:00.146                     
6  M1b  6383           6383  2019-02-18 00:00:02.069                     
8  M1d  6385           6385  2019-02-18 00:00:02.069                     
9  M1e  6386           6386  2019-02-18 00:00:02.069                     

  vitesse longueur  
0                   
2                   
6                   
8                   
9